src/main/java/com/artipie/settings/YamlSettings.java

```diff
@@ -214,106 +168,76 @@ private static CachedUsers initAuth(final YamlMapping settings) {
-                    creds.string("client-id"),
-                    Map.of("secret", creds.string("client-password")),
-                    null
-                )
-            );
-        });
-
-        /**
-         * Transform yaml to completion action with authentication.
-         */
-        private final Function<YamlMapping, Authentication> map;
+        private final YamlMapping cfg;
 
         /**
          * Ctor.
-         *
-         * @param map Transform yaml to completion action with users.
+         * @param cfg Settings config
          */
-        CredentialsType(final Function<YamlMapping, Authentication> map) {
-            this.map = map;
+        public PolicyStorage(final YamlMapping cfg) {
+            this.cfg = cfg;
         }
 
         /**
-         * Get {@code CredentialsType} that corresponds to type from credentials
-         * yaml mapping or {@code ENV} if {@code yaml} is null.
-         *
-         * @param yaml Credentials yaml mapping.
-         * @return CredentialsType.
+         * Read policy storage from config yaml. Normally policy storage should be configured
+         * in `policy` yaml section, but, if policy is absent, storage should be specified in
+         * credentials sections for `artipie` credentials type.
+         * @return Storage if present
          */
-        static CredentialsType valueOf(final YamlMapping yaml) {
-            CredentialsType res = ENV;
-            if (yaml != null) {
-                res = CredentialsType.valueOf(
-                    yaml.string(YamlSettings.NODE_TYPE).toUpperCase(Locale.getDefault())
-                );
+        public Optional<Storage> parse() {
+            Optional<Storage> res = Optional.empty();
+            final YamlSequence credentials = this.cfg.yamlSequence(YamlSettings.NODE_CREDENTIALS);
+            final YamlMapping policy = this.cfg.yamlMapping(YamlSettings.NODE_POLICY);
+            if (credentials != null && !credentials.isEmpty()) {
+                final Optional<YamlMapping> asto = credentials
+                    .values().stream().map(YamlNode::asMapping)
+                    .filter(
+                        node -> YamlSettings.ARTIPIE.equals(node.string(YamlSettings.NODE_TYPE))
+                    ).findFirst().map(node -> node.yamlMapping(YamlSettings.NODE_STORAGE));
+                if (asto.isPresent()) {
+                    res = Optional.of(
+                        new StoragesLoader().newObject(
+                            asto.get().string(YamlSettings.NODE_TYPE),
+                            new Config.YamlStorageConfig(asto.get())
+                        )
+                    );
+                } else if (policy != null
+                    && YamlSettings.ARTIPIE.equals(policy.string(YamlSettings.NODE_TYPE))
+                    && policy.yamlMapping(YamlSettings.NODE_STORAGE) != null) {
+                    res = Optional.of(
+                        new StoragesLoader().newObject(
```

Comment:
`StoragesLoader` problem. Storage factories already initiated for this moment.

Reply:
will be fixed in #1270